File: TwoPointer/424.longest-repeating-character-replacement.py
# @leet start
from collections import defaultdict


class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        l = res = maxfreq = 0
        freq = defaultdict(int)

        for r, c in enumerate(s):
            freq[c] += 1
            maxfreq = max(maxfreq, freq[c])

            while r - l + 1 - maxfreq > k:
                freq[s[l]] -= 1
                l += 1

            res = max(res, r - l + 1)

        return res


# @leet end

File: TwoPointer/424.longest-repeating-character-replacement.go
package leetcode

// @leet start
func characterReplacement(s string, k int) int {
	var (
		freq            = map[byte]int{}
		maxFreq, res, l int
	)
	for r := 0; r < len(s); r++ {
		freq[s[r]]++
		maxFreq = max(maxFreq, freq[s[r]])

		for r-l+1-maxFreq > k { // NOTE: greed
			freq[s[l]]--
			l++
		}
		res = max(res, r-l+1)
	}
	return res
}

// @leet end

